compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -214,6 +213,122 @@ enum BuiltinImplConditions<'tcx> {
+                    let differ_by_single_const_arg = if
+                        substs1.len() == 1
+                        && substs2.len() == 1
+                        && let ty::subst::GenericArgKind::Const(c1) = substs1[0].unpack()
+                        && let ty::subst::GenericArgKind::Const(c2) = substs2[0].unpack()
+                        && let Some(s1) = c1.val().try_to_scalar_int()
+                        && let Some(s2) = c2.val().try_to_scalar_int()
+                        && s1 != s2
+                    {
+                        true
+                    } else {
+                        false
+                    };
+                    !differ_by_single_const_arg
+                }
+            }
+            _ => false,
+        },
+        ty::Str => matches!(k, ty::Str),
+        ty::Array(..) => matches!(k, ty::Array(..)),
+        ty::Slice(..) => matches!(k, ty::Slice(..)),
+        ty::RawPtr(ptr) => matches!(k, ty::RawPtr(ptr2) if ptr.mutbl == ptr2.mutbl),
+        ty::Dynamic(trait_info, ..) => {
+            matches!(k, ty::Dynamic(trait_info2, ..) if
+                trait_info.principal_def_id() == trait_info2.principal_def_id()
+            )
+        }
+        ty::Ref(_, _, mutbl) => matches!(k, ty::Ref(_, _, mutbl2) if mutbl == mutbl2),
+        ty::FnDef(def_id, _) => matches!(k, ty::FnDef(def_id2, _) if def_id == def_id2),
+        ty::Closure(def_id, _) => matches!(k, ty::Closure(def_id2, _) if def_id == def_id2),
+        ty::Generator(def_id, ..) => {
+            matches!(k, ty::Generator(def_id2, ..) if def_id == def_id2)
+        }
+        ty::Never => matches!(k, ty::Never),
+        ty::Tuple(tys) => matches!(k, ty::Tuple(tys2) if tys.len() == tys2.len()),
+        ty::FnPtr(tys) => {
+            matches!(
+                k,
+                ty::FnPtr(tys2)
+                    if tys.skip_binder().inputs().len() == tys2.skip_binder().inputs().len()
+            )
+        }
+        ty::Opaque(def_id, _) => matches!(k, ty::Opaque(def_id2, _) if def_id == def_id2),
+        ty::Foreign(def_id) => matches!(k, ty::Foreign(def_id2) if def_id == def_id2),
+        ty::Param(_) => {
+            // Note, we simplify parameters for the obligation but not the impl
+            // so that we do not reject a blanket impl but do reject more
+            // concrete impls if we're searching for `T: Trait`.
+            matches!(k, ty::Placeholder(..))
+        }
+        ty::Projection(_) => {
+            // When treating `ty::Param` as a placeholder, projections also
+            // don't unify with anything else as long as they are fully normalized.
+            //
+            // We will have to be careful with lazy normalization here.
+            oblig_ty.has_infer_types_or_consts() || matches!(k, ty::Placeholder(..))
+        }
+        ty::Infer(_) => true,
+
+        ty::GeneratorWitness(..) | ty::Placeholder(..) | ty::Bound(..) | ty::Error(_) => {
```

Comment:
surprised that `ty::Placeholder` is unreachable here, that variant should be reachable afaik